dulwich/repo.py

```diff
@@ -676,10 +677,17 @@ def __init__(self, root):
             raise NotGitRepository(
                 "No git repository was found at %(path)s" % dict(path=root)
             )
+        worktree_commondir = os.path.join(self._controldir, COMMONDIR)
+        if(os.path.isfile(worktree_commondir)):
+            with open(worktree_commondir, 'r') as f:
+                self._commondir = os.path.join(self.controldir(),\
+                                                   f.read().rstrip("\n"))
+        else:
+            self._commondir = self._controldir
         self.path = root
-        object_store = DiskObjectStore(os.path.join(self.controldir(),
+        object_store = DiskObjectStore(os.path.join(self.commondir(),
```

Comment:
... Actually I have understood how the `tests/data/repos/` works. I have added the test in c74aa2452aba3fee822595deff18314e2387d29c.